### ISSUE TITLE
ngIf improvements for Rx

### ISSUE THREAD

**Issue by mhevery** [angular/angular#13061]
# Goals

When writing Rx style code it is often necessary to write code like this
```
{{ (userObs|async).?lastName}}, {{ (userObs|async).?firstName}}
```

The issue is that:
1) every `async` creates a listener and 
2) the need to use the `.?` makes coding cumbersome.

What is needed is to be able to subscribe to the observable once and then refer to it in subsequent location. Additionally we need a way to display an alternative template while the observable has not yet yielded a value.

# Proposal

- Allow `ngIf` to assign data to local value.
- Add an `else` feature to `ngIf` so that an alternative template can be shown until the observable yields a value.

```
<div *ngIf="userObs | async; else loading; let user">
  {{ user.lastName}}, {{ user.firstName }}
</div>
<template #loading>
  loading...
</template>
```

NOTE: the microsyntax is consistent with existing implementation of the microsyntax. For expanded form see:

```
<template [ngIf]="userObs | async" [ngIfElse]="loading" let-user>
  {{ user.lastName}}, {{ user.firstName }}
</template>
<template #loading>
  loading...
</template>
```

**Comment by fxck** [comment]
This would be useful in many other places, I wouldn't limit it to ngIf.

**Comment by DzmitryShylovich** [comment]
> Allow ngIf to assign data to local value.

I think it would be better to allow assign result of the async pipe to a local var https://github.com/angular/angular/issues/12237

**Comment by mhevery** [comment]
@DzmitryShylovich 
> I think it would be better to allow assign result of the async pipe to a local var #12237

Pipes are not allowed to do that, and it would not fit with our mental model.

**Comment by thelgevold** [comment]
I am not a huge fan of this proposal as it "encourages" logic in the templates.

IMHO I think this logic should be resolved in code for better testability/reusability.

In these cases I think it's better to use an object with a default value and manual rx subscription over async pipe(s). 

Also, I think it's better to remove the repeated `asyncs` by moving it to a wrapping ngIf rather than putting it on each individual bound field.

**Comment by olegdunkan** [comment]
IMHO It's enough 
````
<template [ngIf]="userObs | async" let-user>
  {{ user.lastName}}, {{ user.firstName }}
</template>
````
I agree with @thelgevold if some directive requires something existing inside templates it source of common pitfalls.
But any way this feature will be very helpful.

**Comment by DzmitryShylovich** [comment]
Then we should do the same with all other structural directives.

**Comment by olegdunkan** [comment]
But what will be if write: 
````
<template [ngIf]="userObs" let-user>
  {{ user.lastName}}, {{ user.firstName }}
</template>
````
Ambiguous thing could happen, will we use 'user' or 'userObs'.  'async' correlates with let-<name_of_variable>it also source of common pitfalls.

**Comment by DzmitryShylovich** [comment]
@olegdunkan `user` will be the same as `userObs`

**Comment by olegdunkan** [comment]
@DzmitryShylovich ok, aka alias) but it will stay ambiguous.

**Comment by kylecordes** [comment]
The notion of a ngIf which captures the value of the expression result and makes it available, has been around a while. I have an implementation of it linked below, but I'm pretty sure I saw another implementation prior to this (about a year ago).

https://github.com/OasisDigital/angular2-firebase/blob/master/ts/ngWhen.ts

It is essentially a blend of ngIf and ngFor, written a long time ago based on a (probably incomplete) understanding of how those things work all the way down.

Somewhere I got the impression that this idea had been considered and rejected, so I'm thrilled to see the core team looking at it again now. Very useful addition.
